File: JS/workbook2/stringGame2.js
"use strict";

const fs = require("fs");

const input =
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString().trim().split("\n")
    : `1
abaaaba
3`.split("\n");

const t = Number(input.shift());
const game = [];

for (let i = 0; i < t; i++) {
  const string = input.shift();
  const k = Number(input.shift());

  game.push([string, k]);
}

const solution = (game) => {
  const [string, k] = game;
  const arr = Array.from(
    Array("z".charCodeAt(0) - "a".charCodeAt(0) + 1),
    () => []
  );
  const chs = string.split("");

  chs.forEach((v, i) => {
    arr[v.charCodeAt(0) - "a".charCodeAt(0)].push(i);
  });

  let min = Infinity - 1;
  let max = -1;

  for (let i = 0; i < arr.length; i++) {
    if (arr[i].length < k) continue;

    for (let j = k - 1; j < arr[i].length; j++) {
      const dist = arr[i][j] - arr[i][j - k + 1];

      min = Math.min(min, dist);
      max = Math.max(max, dist);
    }
  }

  return [min + 1, max + 1];
};

game.forEach((v) => {
  const result = solution(v);

  if (result[0] === Infinity) console.log(-1);
  else console.log(result.join(" "));
});

File: src/workbook2/StringGame2.java
package workbook2;

import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class StringGame2 {
    public static void main(String[] args) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine());
        List<String[]> game = new ArrayList<>();

        for (int i = 0; i < t; i++) {
            String st = br.readLine();
            String k = br.readLine();

            game.add(new String[]{st, k});
        }

        for(String[] g : game){
            int[] re = solution(g);

            if(re[0] == Integer.MAX_VALUE) bw.write("-1\n");
            else bw.write(String.valueOf(re[0]) + " " + String.valueOf(re[1]) + "\n");
        }

        bw.flush();
        bw.close();
    }

    static int[] solution(String[] game){
        String st = game[0];
        int k = Integer.parseInt(game[1]);

        List<Integer>[] arr = new List['z' - 'a' + 1];

        for (int i = 0; i < arr.length; i++) {
            arr[i] = new ArrayList<>();
        }

        String[] chs = st.split("");

        for(int i = 0; i < chs.length; i++){
            char v = chs[i].charAt(0);
            arr[v - 'a'].add(i);
        }

        int min = Integer.MAX_VALUE - 1, max = Integer.MIN_VALUE;

        for (int i = 0; i < arr.length; i++) {
            if(arr[i].size() < k) continue;

            for(int j = k - 1; j < arr[i].size(); j++){
                int dist = arr[i].get(j) - arr[i].get(j - k + 1);

                min = Math.min(min, dist);
                max = Math.max(max, dist);
            }
        }

        return new int[]{min + 1, max + 1};
    }
}
